synapse/module_api/__init__.py

```diff
@@ -748,6 +749,36 @@ def record_user_external_id(
             )
         )
 
+    async def create_login_token(
+        self,
+        user_id: str,
+        duration_in_ms: int = (2 * 60 * 1000),
+        auth_provider_id: Optional[str] = None,
+        auth_provider_session_id: Optional[str] = None,
+    ) -> str:
+        """Create a login token suitable for m.login.token authentication
+
+        Added in Synapse v1.69.0.
+
+        Args:
+            user_id: gives the ID of the user that the token is for
+
+            duration_in_ms: the time that the token will be valid for
+
+            auth_provider_id: the ID of the SSO IdP that the user used to authenticate
+                to get this token, if any. This is encoded in the token so that
+                /login can report stats on number of successful logins by IdP.
+
+            auth_provider_session_id: The session ID got during login from the SSO IdP,
+                if any.
+        """
+        return self._hs.get_macaroon_generator().generate_short_term_login_token(
+            user_id,
+            auth_provider_id or "",
```

Comment:
If we're defaulting to `""`, why not make it the default value of the `auth_provider_id` parameter?

Reply:
The reason we're defaulting to `""` is because of how we're building the macaroon. Since the plan is to save the login tokens in database, it will become a properly nullable field, hence the change here.

Reply:
Right I see, thanks.

Reply:
Right I see, thanks. Could you add a comment explaining this then? Otherwise it's pretty confusing to have the definition of `auth_provider_id` not default to the empty string despite it being the default value for it in practice.

Reply:
Just did that.